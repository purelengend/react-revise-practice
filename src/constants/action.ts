export const URL_ACTION = {
  SORT: "SORT",
  ORDER: "ORDER",
  PAGE: "PAGE",
  FILTER: "FILTER",
} as const;
