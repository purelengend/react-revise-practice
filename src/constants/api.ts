export const API = {
  STUDENT: "VITE_STUDENT_API",
  IMAGE: "VITE_UPLOAD_IMAGE_API",
  IMGBB_KEY: "VITE_IMGBB_API_KEY",
} as const;
