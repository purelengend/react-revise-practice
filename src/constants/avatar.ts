export const DEFAULT_AVATAR_URL = "/assets/images/user-photo.svg";
