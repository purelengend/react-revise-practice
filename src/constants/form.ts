export const FORM_TITLE = {
  ADD: "Add student",
  EDIT: "Edit student",
  DELETE: (id: string) =>
    `Are you sure to delete student with enroll number: ${id}?`,
} as const;

export const FORM_ERROR = {
  NAME: "Your name must have at least 4 characters",
  EMAIL: "Please enter a valid email address, e.g., abc@mail.com",
  PHONE: "Please enter a valid phone number with 10 digits",
} as const;
