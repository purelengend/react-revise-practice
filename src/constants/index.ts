export * from "./axios";
export * from "./query";
export * from "./role";
export * from "./avatar";
