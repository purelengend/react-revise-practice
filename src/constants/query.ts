export const ROUTES = {
  STUDENT: "students",
} as const;

export const QUERY_KEY = {
  STUDENT: "students",
  STUDENT_COUNT: "student_count",
} as const;

export const QUERY_PARAMS = {
  PAGE: "page",
  LIMIT: "limit",
  SORTBY: "sortBy",
  ORDER: "order",
  NAME: "name",
} as const;
