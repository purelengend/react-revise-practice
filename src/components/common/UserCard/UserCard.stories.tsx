import type { Meta, StoryObj } from "@storybook/react";

// Components
import UserCard from ".";

// Constants
import { DEFAULT_AVATAR_URL } from "@/constants";

const meta: Meta<typeof UserCard> = {
  title: "Components/UserCard",
  component: UserCard,
  argTypes: {
    name: {
      description: "The name of the user card",
    },
    role: {
      description: "The role of the user card",
    },
    avatarUrl: {
      description: "The avatar url of the user card",
    },
  },
  parameters: {
    controls: {
      expanded: true,
    },
  },
};

export default meta;

type Story = StoryObj<typeof UserCard>;

export const Primary: Story = {
  args: {
    name: "John",
    role: "Admin",
    avatarUrl: DEFAULT_AVATAR_URL,
  },
};
