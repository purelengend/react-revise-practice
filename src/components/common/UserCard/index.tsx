import {
  Card,
  CardBody,
  CardProps,
  Center,
  Heading,
  Image,
  Text,
} from "@chakra-ui/react";
import { memo } from "react";

// Types
import { Role } from "@/types";

// Constants
import { DEFAULT_AVATAR_URL } from "@/constants";

export type UserCardProps = CardProps & {
  name: string;
  role: Role;
  avatarUrl?: string;
};
const UserCard = memo(
  ({ name, role, avatarUrl = DEFAULT_AVATAR_URL, ...rest }: UserCardProps) => {
    return (
      <Card {...rest} bg="transparent" variant="unstyled">
        <CardBody as={Center} flexDirection="column">
          <Image
            boxSize={32}
            src={avatarUrl}
            borderRadius="50%"
            objectFit="contain"
          />
          <Heading fontWeight="bold" textAlign="center" mt={5} fontSize="md">
            {name}
          </Heading>

          <Text
            mt={2.5}
            textAlign="center"
            fontWeight="500"
            fontSize="sm"
            color="yellow.200"
          >
            {role}
          </Text>
        </CardBody>
      </Card>
    );
  },
);

export default UserCard;
