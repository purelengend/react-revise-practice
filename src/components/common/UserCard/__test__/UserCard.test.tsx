import { render, screen } from "@testing-library/react";
import UserCard, { UserCardProps } from "..";

// Constants
import { DEFAULT_AVATAR_URL, ROLES } from "@/constants";

describe("UserCard test cases", () => {
  const mockUserCardProps: UserCardProps = {
    name: "test name",
    role: ROLES.ADMIN,
    avatarUrl: "test avatarUrl",
  };
  it("should render correctly", () => {
    const { container } = render(<UserCard {...mockUserCardProps} />);
    expect(container).toMatchSnapshot();
  });

  it("should render with default avatar URL", () => {
    render(
      <UserCard name={mockUserCardProps.name} role={mockUserCardProps.role} />,
    );

    const avatarElement = screen.getByRole<HTMLImageElement>("img");

    const hasDefaultAvatarUrl = avatarElement.src.includes(DEFAULT_AVATAR_URL);

    expect(hasDefaultAvatarUrl).toEqual(true);
  });
});
