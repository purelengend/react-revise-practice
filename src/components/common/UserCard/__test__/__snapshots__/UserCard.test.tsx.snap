// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserCard test cases should render correctly 1`] = `
<div>
  <div
    class="chakra-card css-qdyteg"
  >
    <div
      class="chakra-card__body css-xe2qne"
    >
      <img
        alt="avatar-of-test name"
        class="chakra-image css-161novc"
        src="test avatarUrl"
      />
      <h2
        class="chakra-heading css-1j31z2v"
      >
        test name
      </h2>
      <p
        class="chakra-text css-fmlsi3"
      >
        Admin
      </p>
    </div>
  </div>
</div>
`;
